### PR TITLE
Add charges intermediate model

### PR DESCRIPTION
**Pull request by ifoukarakis**
#### Summary

Intermediate model that denormalizes on prem subscriptions on the charge level.

- [x] Add extra logic for reconciling bug where license start date is 0 and license end date is <> 0.
- [x] Merge on charge level, following existing opportunity creation logic.
- [x] Classify as renewal, expansion, contraction.
- [x] Tidy staging subscription model. 
- [ ] Compare against current data.

### CODE REVIEW

File: transform/mattermost-analytics/models/intermediate/sales/int_charges_filtered_onprem.sql [hunk 1]
```diff
@@ -0,0 +1,50 @@
+{{config({
```
Comment (catalintomai) [inline transform/mattermost-analytics/models/intermediate/sales/int_charges_filtered_onprem.sql:1]: We seemingly are using onprem (here) and self-hosted (see https://mattermost.atlassian.net/wiki/spaces/DATAENG/pages/2435874817/Self-serve) interchangeably. Wonder if we should decide and align on one one term to use across.

File: transform/mattermost-analytics/models/staging/stripe/stg_stripe__subscriptions.sql [hunk 2]
```diff
@@ -7,53 +7,82 @@ with source as (
...
-        plan:"name"::varchar as edition,
         plan:"product"::varchar as product_id,
-        TO_TIMESTAMP_NTZ(metadata:"cws-date-converted-to-paid"::int) as converted_to_paid_at,
-        TO_TIMESTAMP_NTZ(metadata:"cws-license-start-date"::int) as license_start_at,
-        TO_TIMESTAMP_NTZ(metadata:"cws-license-end-date"::int) as license_end_at,
-        TO_TIMESTAMP_NTZ(metadata:"cws-actual-renewal-date"::int / 1000) as actual_renewal_at,
+
+        -- Subscription data
+        created as created_at,
+        current_period_end as current_period_end_at,
```
Comment (catalintomai) [inline transform/mattermost-analytics/models/staging/stripe/stg_stripe__subscriptions.sql:24]: nit: flipping these two.

File: transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql [hunk 3]
```diff
@@ -27,14 +17,16 @@ where
     and p.name <> 'Premier Support'
     -- Skip incomplete subscriptions
     and s.status <> 'incomplete_expired'
+    -- Ignoring subscriptions that come from admin portal/sales
+    and s.renewal_type is null
```
Comment (catalintomai) [inline transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql:21]: Is "subscriptions that come from admin portal/sales" == "renewal_type is null" ?

File: transform/mattermost-analytics/models/intermediate/sales/int_charges_filtered_onprem.sql [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    s.actual_renewal_at,
+    ili.quantity,
+    ili.quantity - LAG(ili.quantity) OVER (PARTITION BY s.customer_id ORDER BY i.created_at) as seat_difference,
+    datediff(day, lag(license_end_at) OVER (PARTITION BY s.customer_id ORDER BY i.created_at), license_start_at) as days_since_previous_license_end,
+    row_number() over(partition by s.customer_id order by i.created_at) as charge_order,
+    case
+        when s.renewed_from_subscription_id is not null then true
+        when days_since_previous_license_end < 60 then true
+        else false end
+    as is_renewal,
```
Comment (catalintomai) [inline transform/mattermost-analytics/models/intermediate/sales/int_charges_filtered_onprem.sql:28]: @johndavidlugtu to help validate/confirm the renewal logic?

**Comment by catalintomai** [comment]
@sri-byte , please make sure to review the PR. Thanks.

File: transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql (hunk 3 above)
Comment (ifoukarakis) [reply transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql:21]: > Is "subscriptions that come from admin portal/sales" == "renewal_type is null" or "renewal_type <> 'sales-only'"? Open question if/why "admin portal == sales_only".

According to `Self-serve` document:

> When a renewal type is sales-only, that subscription cannot be renewed self serve, and it doesn’t necessarily fit into the criteria of this document.

The values are either `sales-only` or `null`. It's expressed  this way here in order to explicit about this expectation.

File: transform/mattermost-analytics/models/intermediate/sales/int_charges_filtered_onprem.sql (hunk 4 above)
Comment (ifoukarakis) [reply transform/mattermost-analytics/models/intermediate/sales/int_charges_filtered_onprem.sql:28]: I'm also still working on that. Seems that there's some additional cases (i.e. trials) that we should consider.

File: transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql (hunk 3 above)
Comment (catalintomai) [reply transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql:21]: Makes sense, could be useful to add a link to the documentation/Confluence page for future reference.

Comment (sri-byte) [reply transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql:22]: Adding Nick's comment here - The easiest way to determine if something was a sales-serve subscription vs not is by looking for a SUBSCRIPTIONS.METADATA key cws-purchase-source and value admin. Any subscription with that value was done so sales-serve.

**Review by sri-byte — APPROVED** [review]

File: transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql (hunk 3 above)
Comment (catalintomai) [reply transform/mattermost-analytics/tests/staging/stripe/test_subscription__onprem_has_license_data.sql:22]: Thanks, added it to the Confluence page. Still not completely clear from his comment if ALL the sales-serve subscriptions follow this rule.

File: transform/mattermost-analytics/models/staging/stripe/stg_stripe__subscriptions.sql [hunk 5]
```diff
@@ -7,58 +7,86 @@ with source as (
...
+        metadata:"cws-installation-state"::varchar as cws_installation_state,
+
+        -- License specific data
+        TO_TIMESTAMP_NTZ(metadata:"sfdc-original-start-date"::varchar) as sfdc_migrated_started_at,
+        TO_TIMESTAMP_NTZ(metadata:"cws-actual-renewal-date"::int / 1000) as actual_renewal_at,
+        case
+            -- Handle backfills
+            when sfdc_migrated_license_id is not null then sfdc_migrated_started_at
+            -- Handle bug whenever license start date doesn't exist but end date exists
+            when metadata:"cws-license-start-date"::int = 0 and metadata:"cws-license-end-date"::int > 0 then TIMEADD(year, -1, TO_TIMESTAMP_NTZ(metadata:"cws-license-end-date"::int))
```
Comment (catalintomai) [inline transform/mattermost-analytics/models/staging/stripe/stg_stripe__subscriptions.sql:62]: Wondering if it is an opportunity to align this logic with the logic in arr_transactions (that relies on ORGM data and uses fixed values like '2022-03-31') cc: @RoyI99

**Comment by ifoukarakis** [comment]
Closing as work has changed scope.

**Pull request closed without merging**